### DISCUSSION TITLE
New grain formation using Grain tracker

### DISCUSSION THREAD

**Discussion by ritamcj** [idaholab/moose#19191] [Q&A Modules: Phase field] [answered]
Dear Moose users,
I am trying to simulate nucleation in a polycrystal using Grain tracker. I came across the option "reserved order parameters" in Grain tracker but I am unable to figure out how to invoke new grain formation using this. I tried various strategies as follows:
1. I used discretenucleation to increase new grain order parameters (say gr3,gr4,gr5) to values greater than the threshold required to recognise a new feature. I kept op_num = 6, initial grains = 3 and used polycrystalvariables and polycrystalkernel to evolve grain op's with prefix "gr". 
  ```
[./nucleation_energy]
    type = DiscreteNucleation
    f_name = Fn
    op_names  = 'gr0 gr1 gr2 gr3 gr4 gr5'
    op_values = '0.1 0.1 0.1 1 1 1'
    penalty = 20
    penalty_mode = MATCH
    map = map
    block = 'face1_QUAD4  face2_QUAD4  face3_QUAD4'
    outputs = exodus
  [../]
```

2. Increase the number of order parameters "op_num" in my input file by writing a simple function. Something like:
```
[GlobalParams]
  op_num = op_evolution_function
  deformed_grain_num = 3
  var_name_base = gr
  grain_tracker = grain_tracker
[../]

[Functions]
  [./op_evolution_function]
    type = ParsedFunction
    value = deformed_grain_num + 1
  [../]
[]
```
This did not work as the syntax to read a function value is likely incorrect. I have come across functions in Grain tracker such as newGrainCreated, getNewGrainIDs but am unable to figure out **how to invoke nucleation through new grain order parameters**. I would really appreciate guidance on possible strategies to tackle this problem.

Warm Regards,
Ritam

**Comment by laagesen** [comment, marked as answer]
Hi Ritam,

If you haven't already, please have a look at the documentation page

https://mooseframework.inl.gov/modules/phase_field/Nucleation/DiscreteNucleation.html

and especially the section Direct Order Parameter Modification. You will need to add a `DiscreteNucleationForce` and `Reaction` kernel that act on the reserved order parameter. You shouldn't need to do either of the other things you tried, but you will need a map and inserter to inform the `DiscreteNucleationForce` kernel.

**Reply by dschwen** [reply]
Also make sure you run nucleation only on the reserved order parameters. Those are not assigned any grains by the grain tracker and can be use to stage new nuclei.

**Comment by ritamcj** [comment]
Hi Larry and Daniel,
Thank you so much for your suggestions. I am able to see nucleating grains individually through their order parameters (say gr9). I am now trying to observe the evolving grain structure, but the new nuclei aren't observed in "unique_grains" in GrainTracker even though the new grain order parameters exceed the threshold. However, the halos do update. Any pointers regarding this? I have applied a grain growth model through "polycrystalstoredenergy" and am able to observe grain growth in pre-existing grains.
Ritam


https://user-images.githubusercontent.com/81842276/139257266-69cd1723-2b71-4260-81f9-43055feb46fe.mp4


https://user-images.githubusercontent.com/81842276/139257277-19c7af51-5a77-444d-9f68-c6245435f8d3.mp4


https://user-images.githubusercontent.com/81842276/139258328-539df671-c809-48a1-a5cb-c7f45af79e0b.mp4

**Comment by laagesen** [comment]
Ritam, you will need to use a significantly higher mesh resolution than shown here. The GrainTracker is probably having a hard time distinguishing new grains from existing grains. The nucleation rate also seems to be very high so it may be having a hard time distinguishing between individual nucleation events. Please make sure you have at least 3 mesh elements through the interfaces between order parameters prior to trying anything with nucleation. Once you observe grain growth under these conditions, you can then add in nucleation.

**Reply by ritamcj** [reply]
Hi Larry,
I am happy to inform you that I am able to visualize the evolving microstructure with nucleating grains, which grow further with stored deformation energy based on dislocation density evolution. I used a very fine mesh and a slower nucleation rate and it worked. Many thanks for your suggestions.
Warm regards,
Ritam


https://user-images.githubusercontent.com/81842276/139530993-2ff7a244-1a82-4fa2-bd95-2de5d5f8fbb5.mp4


https://user-images.githubusercontent.com/81842276/139530997-d83b2e11-3550-4d56-9320-39e125a6786b.mp4

**Reply by laagesen** [reply]
Great! Glad that worked.

**Comment by Sabalei** [comment]
Hi Ritam,
I'm trying to simulate static recrystallization. I also applied a grain growth model through "polycrystalstoredenergy". 
I have inserted the nucleus through Direct order parameter modification. However, new nuclei shrink after formation rather than grow further.  I tried to increase the hold time and Dislocation density, but it didn't solve the problem. Any pointers regarding this?

**Reply by ritamcj** [reply]
Hi Sabalei,
You can increase the order modification parameter (v1) to increase the driving force for growth of nuclei. Please see https://mooseframework.inl.gov/source/kernels/DiscreteNucleationForce.html
Now, you should be able to see growth followed by shrinkage and annihilation of nuclei.
Warm regards,
Ritam

**Reply by laagesen** [reply]
Thanks @ritamcj!

**Reply by mdh23rct** [reply]
@ritamcj @Sabalei @GiudGiud Dear all,
Have you got any grain coalescence problem in your new grain formation simulation? I am finding two grains with the same grain order parameter are coalescing with each other (Please see the attached image). However, I would like to prevent the coalescence of grains and keep them sepearate with a grain boundary. Would you be kind to let me know what sort of parameters I should use in the GrainTracker userobject to make that happen? Your advice will be highly appreciated. Thanks.
![grain-coalescence](https://github.com/user-attachments/assets/de00f4f9-62f8-461e-9b96-89147d1ffc47)

**Reply by Sabalei** [reply]
Are these new grains formed by nucleation? Whether remap is enabled? Coalescence can be prevented  by remap two grains that are very close together into different order parameters.

**Reply by mdh23rct** [reply]
@Sabalei @GiudGiud 
Yes, new grains are forming by discrete nucleation. As you suggested, I have enabled the remapping. However, I am getting a 
new problem. Duplicate grains are forming at the same location when remap_grains = true. Please see the attached image. Any solution for that? Thanks.
![Duplicate grain-nucleation-1](https://github.com/user-attachments/assets/25a648aa-7e9b-4cc0-8e19-8234aae1e58d)

**Reply by Sabalei** [reply]
I think you need to modify the Useobject DiscreteNucleationInserter code makes the distance between new grains is greater than a set value.
